### PR TITLE
[[FEAT]] MongoDB Runner & Lifecycle events

### PR DESCRIPTION
**Pull request by jmendiara**
Add supports for node4

BREAKING CHANGES:
- Drop Config Runner in favour of Startup.check() generic method
- Drop CompositeRunner

### CODE REVIEW

File: src/mongo.ts [hunk 1]
```diff
@@ -19,29 +19,143 @@ import Therror, { Classes }  from 'therror';
...
   uri: string;
   options: MongoClientOptions;
 
-  constructor(opts: MongoRunnerOptions) {
-    this.client = opts.client;
-    this.uri = opts.uri;
-    this.options = opts.options;
+  // MongoDriver reconnects every second
+  protected startupRetryInterval = 1000;
+  protected startupReconnectTries: number;
```
Comment (palmerabollo) [inline src/mongo.ts]: Remove vars

File: src/mongo.ts [hunk 2]
```diff
@@ -19,29 +19,143 @@ import Therror, { Classes }  from 'therror';
...
+
+    this.options.server.reconnectTries = typeof this.options.server.reconnectTries === 'undefined' ?
+      -1 : // Number of retries
+      this.options.server.reconnectTries;
+
+    this.options.server.reconnectInterval = typeof this.options.server.reconnectInterval === 'undefined' ?
+      1000 : // Time between retries
+      this.options.server.reconnectInterval;
+
+    this.options.db.bufferMaxEntries = typeof this.options.db.bufferMaxEntries === 'undefined' ?
```
Comment (palmerabollo) [inline src/mongo.ts]: typeof null === 'undefined'
false

File: src/precon.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+/** A configuration service must have a `validate()` method that throws when the validation didn't succeed */
+export interface PreconRunnerConfig {
+  preconditions: Precondition<any>[];
+}
+
+/**
+ * Validates an app configuration. It merely be used as a pre-condition check to other runners or services
+ */
+export class PreconRunner extends Runner<void> {
```
Comment (palmerabollo) [inline src/precon.ts]: I don't like abbreviations such as "Precon" (vs Precondition)

File: src/runner.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+export abstract class Runner<T> extends EventEmitter implements Runner<T> {
+  constructor(public name: string) {
+    super();
+  }
+
+  protected abstract doStart(): Promise<T>
+  protected abstract doStop(): Promise<T>
+
+  public info(): any { ; };
```
Comment (palmerabollo) [inline src/runner.ts]: { ; } vaya mierda

File: src/precon.ts (hunk 3 above)
Comment (jmendiara) [inline src/precon.ts]: done

File: src/mongo.ts (hunk 2 above)
Comment (jmendiara) [inline src/mongo.ts]: done

File: src/runner.ts (hunk 4 above)
Comment (jmendiara) [inline src/runner.ts]: I know... TS...

**Pull request merged**